---
title: "Analysis - Querying Albutein Transcriptomics Dataset against LINCS2 Database"
author: "Nhi Hin"
date: "2023-06-07"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Load Packages

```{r message=FALSE, warning=FALSE}
# Run under R 4.2.2
library(Seurat)
library(dplyr)
library(magrittr)
library(readr)
library(tibble)
library(ggplot2)
library(SingleCellExperiment)
library(scater)
library(here)
library(rhdf5)
library(signatureSearch)
library(signatureSearchData)
library(HDF5Array)
library(ExperimentHub)
library(disgenet2r)
Sys.setenv(DISGENET_API_KEY = "030b2e02ff356ed135085de8c366cbbacb3bf9a4")
```


## Objective

- Load in the Albutein (CNS 10X transcriptomics) dataset, separate it into cell types, and run the Endothelial cells' Albutein vs Vehicle comparison as a query against LINCS2 using the CMap methodology as implemented in *signatureSearch*. 

- Results will be exported and also interpreted as follows:

1. Over representation analysis of MOAs and gene targets
2. Network visualizations


## Output Directory

```{r}
output_dir <- here("output", "_queryAlbutein_against_dbLINCS2")
ifelse(!dir.exists(output_dir),
       dir.create(output_dir,
                  recursive = TRUE), 
       FALSE)
```


## Import data

- See `code/locating_albutein_datasets.R` for details on how the below Seurat object was prepared. 

```{r}
seurat_path <- here("data/albutein_seurat.rds")
seurat <- readRDS(seurat_path)
```

## Load Helper functions

```{r}
source(here("code", "helper_functions.R"))
```

## Run Analysis

```{r}
seurat <- patchSeuratMetadataAge(seurat)
```

```{r}
# Note first step takes ~20 mins to run + up to 150Gb in memory 
# depending on dataset due to filtering step
sce <- convertSeuratToSCE(seurat)
gc()

pseudobulk_mat <- aggregateSCEIntoPseudobulkMatrix(sce)
pseudobulk_mat_l <- splitAggregatedMatrixPerCelltype(pseudobulk_mat)

colData(sce) %>% 
  as.data.frame %>%
  dplyr::group_by(treatment_label, sample_ID) %>%
  dplyr::summarise(n=n()) %>% 
  write_csv(file.path(output_dir,
                      "cells_per_pseudobulk_sample.csv"))

zscores_l <- splitPerCelltypeAggregatedMatrixByGroup(
  sce, 
  pseudobulk_mat_l,
  grouping_column = "treatment_label",
  pseudobulk_grouping_column = "sample_ID"
  )
```

### Save / Load Z-Scores

```{r eval=FALSE}
zscores_l %>% saveRDS(file.path(output_dir, "zscores_l.rds"))
zscores_l <- readRDS(file.path(output_dir, "zscores_l.rds"))
```


## Convert mouse ensembl ID to human entrez ID

- Because CMap is in the form of human entrez IDs, we will need to convert the mouse Ensembl IDs in order to do queries. 

- There are some many:1 and potentially 1:many mappings going from mouse to human. In this case we will use the gene expression value with the maximum absolute value (regardless of sign). Other potential options worth trying are "sum" and "mean" (not done yet). 

```{r}
# Takes 1-2 mins per cell type
# Only doing Endothelial for this analysis 
# as CMap doesn't really have any cell lines that closely 
# correspond to any other cell types present in CNS. 
celltypesToKeep <- c("endothelial")

zscores_entrez <- zscores_l[celltypesToKeep] %>%
  lapply(function(x){
    x %>% lapply(function(y){
      df <- y %>%
        as.data.frame %>%
        set_colnames(c("zscore")) %>%
        convertMouseEnsemblToHumanEntrezID(
          functionForReplicates = "absmax")
      ls <- df$zscore %>%
        magrittr::set_names(rownames(df))
        return(ls)
    })
  })
```

```{r eval=FALSE}
zscores_entrez %>% saveRDS(file.path(output_dir, "zscores_df.rds"))
zscores_entrez <-readRDS(file.path(output_dir, "zscores_df.rds"))
```

## Calculate difference of z-scores  

- The idea here is to get a score corresponding to how DE each gene is in a particular comparison. I think we will probably use an actual DE metric for this eventually (for any experiments with n=3), but I can see this method being potentially useful for any experiments with n<3 (the original CMap was prepared in this way as there were only n=1 or n=2 etc.)

```{r}
zscores_entrez$endothelial %>% names

# Get them into a clean format without special chars
for(celltype in names(zscores_entrez)) {
  names(zscores_entrez[[celltype]]) %<>%
    gsub(x = ., pattern = " |-", replacement = "")
}

diffs <- list()
for(celltype in names(zscores_entrez)){
  diffs[[celltype]] <-
    zscores_entrez[[celltype]]$AlbuteinA20024 - 
    zscores_entrez[[celltype]]$VehicleALK53760
}
```

### Histogram of the z-score distribution

```{r}
diffs$endothelial %>% hist(breaks = 100)
```


## Get top and bottom 150 genes based on ranking of difference of z-scores

```{r}
upsets <- list()
downsets <- list()

for(celltype in names(diffs)){
  upsets[[celltype]] <- top_n_names_from_list(diffs[[celltype]],
                                   direction = "top",
                                   n = 150)
  
  downsets[[celltype]] <- top_n_names_from_list(diffs[[celltype]],
                                   direction = "bottom",
                                   n = 150)
}
```

## Top nDE genes 

```{r}
upsets2 <- list()
downsets2 <- list()

for(celltype in names(diffs)){
  upsets2[[celltype]] <- top_n_names_from_list(diffs[[celltype]],
                                   direction = "top",
                                   n = 1688)
  
  downsets2[[celltype]] <- top_n_names_from_list(diffs[[celltype]],
                                   direction = "bottom",
                                   n = 2667)
}

upsets2$test <- upsets2$endothelial
downsets2$test <- downsets2$endothelial
```

### Genes in common with bulk experiment

```{r}
up_genes_in_common <- readRDS(here("output", "_scratch_albutein_comparison", "up_genes_in_common.rds"))

down_genes_in_common <- readRDS(here("output", "_scratch_albutein_comparison", "down_genes_in_common.rds"))

upset_in_common <-
  up_genes_in_common %>% 
  dplyr::arrange(desc(sc_alb_vs_veh)) %>%
  dplyr::pull(entrezid)

downset_in_common <-
  down_genes_in_common %>% 
  dplyr::arrange(desc(sc_alb_vs_veh)) %>%
  dplyr::pull(entrezid)

qsigs$endothelial_in_common <- qSig(
    query = list(upset = upset_in_common,
                 downset = downset_in_common),
    gess_method = "LINCS",
    refdb = lincs2
  )

queries_against_lincs2$endothelial_in_common <-
    gess_lincs(
      qSig = qsigs$endothelial_in_common,
      sortby = "NCS",
      tau = TRUE,
      cmp_annot_tb = lincs_pert_info2,
      by = "pert_id",
      cmp_name_col = "pert_iname",
      workers = 2
    )
```



## Query LINCS2

### Load LINCS2 Database

```{r}
# eh <- ExperimentHub()
# lincs2 <- eh[["EH7297"]]
# saveRDS(lincs2, here("data", "lincs2.rds"))

lincs2 <- readRDS(here("data", "lincs2.rds"))
```

### Perform Queries using SignatureSearch

```{r}
qsigs <- list()
queries_against_lincs2 <- list()

qsigs <- readRDS(file.path(
  output_dir, "qsigs.rds"
))

for (celltype in names(upsets)) {
  qsigs[[celltype]] <- qSig(
    query = list(upset = upsets[[celltype]],
                 downset = downsets[[celltype]]),
    gess_method = "LINCS",
    refdb = lincs2
  )
  
  queries_against_lincs2[[celltype]] <-
    gess_lincs(
      qSig = qsigs[[celltype]],
      sortby = "NCS",
      tau = TRUE,
      cmp_annot_tb = lincs_pert_info2,
      by = "pert_id",
      cmp_name_col = "pert_iname",
      workers = 2
    )
}

```




### Query again but using DE genes

- We will use the DE genes that are FDR < 0.05 and abs(log2 FC) > 0.1. 

```{r}
degs <- "/mnt/home/alkahest.com/nvu/Documents/Analyses/alkahest/sci_app/PPF_vs_Albutein/2022-08-05/seurat/DEG_1.00_ranked/Albutein_against_Vehicle_ALK-53760_all_cells/Major_celltype/endothelial/endothelial_8077_Albutein A-20024_33851_Vehicle ALK-53760.tsv" %>%
  readr::read_tsv()

nrow(degs)
degs %<>%
  dplyr::filter(p_val_adj < 0.05)

degs_up <- degs %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::arrange(desc(avg_log2FC))
nrow(degs_up)
degs_up$avg_log2FC %>% summary()

degs_up %<>%
  dplyr::filter(avg_log2FC > 0.1) 

degs_up %<>%
  dplyr::select(avg_log2FC, gene_id) %>%
  tibble::column_to_rownames("gene_id") %>%
  convertMouseEnsemblToHumanEntrezID(
    functionForReplicates = "absmax"
  ) %>%
  dplyr::arrange(desc(avg_log2FC))

dim(degs_up)

degs_down <- degs %>%
  dplyr::filter(avg_log2FC < 0) %>%
  dplyr::arrange(avg_log2FC)
nrow(degs_down)
degs_down$avg_log2FC %>% summary()

degs_down %<>%
  dplyr::filter(abs(avg_log2FC) > 0.1)

degs_down %<>%
  dplyr::select(avg_log2FC, gene_id) %>%
  dplyr::mutate(avg_log2FC = abs(avg_log2FC)) %>%
  tibble::column_to_rownames("gene_id") %>%
  convertMouseEnsemblToHumanEntrezID(
    functionForReplicates = "absmax"
  ) %>%
  dplyr::arrange(desc(avg_log2FC))
degs_down %>% View
degs_down %>% dim
degs_upset <- degs_up %>% rownames()
degs_downset <- degs_down %>% rownames()
```

```{r}
qsigs$endothelial_degs <- qSig(
  query = list(
    upset = degs_upset,
    downset = degs_downset),
  gess_method = "LINCS",
  refdb = lincs2
)

queries_against_lincs2$endothelial_degs <- 
  gess_lincs(
    qSig = qsigs$endothelial_degs,
      sortby = "NCS",
      tau = TRUE,
      cmp_annot_tb = lincs_pert_info2,
      by = "pert_id",
      cmp_name_col = "pert_iname",
      workers = 2
  )

queries_against_lincs2 %>%
  saveRDS(file.path(output_dir,
               "queries_against_lincs2.rds"))

```



### Save out results / read results

```{r eval=FALSE}
queries_against_lincs2 <- 
  readRDS(file.path(output_dir,
               "queries_against_lincs2.rds"))

queries_against_lincs2_v2 <- 
  readRDS(file.path(output_dir,
               "queries_against_lincs2_v2.rds"))

queries_against_lincs2_v3 <- 
  readRDS(file.path(output_dir,
               "queries_against_lincs2_v3.rds"))

queries_against_lincs2 %>% 
  lapply(result) %>%
  openxlsx::write.xlsx(
    file.path(output_dir,"queries_against_lincs2.xlsx")
    )

tmp <- queries_against_lincs2$endothelial
tmp2 <- queries_against_lincs2$endothelial_degs
tmp3 <- queries_against_lincs2_v2$test
tmp4 <- queries_against_lincs2_v3$endothelial_in_common
queries_against_lincs2$endothelial <- list(
  albutein_vs_veh = tmp,
  albutein_vs_veh_degs = tmp2,
  albutein_vs_veh_v2 = tmp3,
  albutein_vs_veh_incommon = tmp4
)

queries_against_lincs2 %>% saveRDS(
  file.path(output_dir, "queries_against_lincs2_all.rds")
)
```

### Apply Significance Threshold

```{r}
for(celltype in names(queries_against_lincs2)){
  for(comparison in names(queries_against_lincs2[[celltype]])){
    queries_against_lincs2[[celltype]][[comparison]] %<>%
      result %>%
      #dplyr::filter(!is.na(t_gn_sym)) %>%
      dplyr::mutate(sig =   (trend == "up" &
                           WTCS_FDR < 0.05 &
                           NCSct > 0.8)) %>%
  dplyr::distinct(.keep_all=TRUE)
  }
}
```

### Combine Results into Single Dataframe

- Note: The Significance Threshold defined below is arbitrary. May need some reworking/further investigation. 

```{r eval=FALSE}
full_table <- queries_against_lincs2 %>%
  lapply(result) %>%
  bind_rows(.id = "query_celltype") %>%
  #dplyr::filter(!is.na(t_gn_sym)) %>%
  dplyr::mutate(sig =   (trend == "up" &
                           WTCS_FDR < 0.05 &
                           NCSct > 0.8)) %>%
  dplyr::distinct(.keep_all=TRUE) %>%
  dplyr::filter(query_celltype == "endothelial")

```

```{r eval=FALSE}
full_table %>% saveRDS(file.path(output_dir, "full_table.rds"))
```

```{r}
full_table %>%
  dplyr::mutate(hasMOA = !is.na(MOAss)) %>%
  dplyr::filter(sig == TRUE) %>%
  dplyr::group_by(cell, hasMOA) %>% 
  dplyr::summarise(n=n() ) %>% 
  dplyr::filter(hasMOA == TRUE) %>% 
  .$n %>% sum

full_table %>%
  dplyr::distinct(pert,.keep_all=T) %>%
  dplyr::mutate(hasMOA = !is.na(MOAss)) %>%
  dplyr::group_by(hasMOA) %>%
  dplyr::summarise(n=n())
```


### Summary of Results

### How many significant peturbations?

```{r}
# Across all cell lines
queries_against_lincs2$endothelial %>% 
  lapply(function(x){
    x$sig %>% table
  })

# In HUVEC cell line only
queries_against_lincs2$endothelial %>% 
  lapply(function(x){
    x %>%
      dplyr::filter(cell == "HUVEC") %>% .$sig %>% table
  })
```

- Overlap in results from HUVEC cell line

```{r}
queries_against_lincs2$endothelial %>% 
  .[c("albutein_vs_veh", "albutein_vs_veh_v2")] %>%
  lapply(function(x){
    x %>%
      dplyr::filter(
        #cell == "HUVEC",
                    sig == TRUE) %>% 
      dplyr::select(pert_iname, cell,
                    starts_with("WTCS"),
                    NCS,
                    NCSct,
                    t_gn_sym,
                    MOAss)
  }) %>%
  purrr::reduce(function(x,y){
    dplyr::inner_join(x,y,by = c("cell","pert_iname",
                                 "t_gn_sym",
                                 "MOAss"))
  }) %>% dim
```

```{r}
intersect(degs_upset, upsets2$test) %>% length
intersect(degs_downset, downsets2$test) %>% length
```

```{r}
huvec_sig_results <- queries_against_lincs2$endothelial %>%
  lapply(function(x){
    x %>%
      dplyr::filter(cell == "HUVEC") %>%
      dplyr::filter(sig == TRUE) %>%
      dplyr::select(pert_iname, 
                    starts_with("NCS"),
                    starts_with("WTCS"),
                    t_gn_sym)
  })

intersect(huvec_sig_results$albutein_vs_veh_degs$pert_iname,huvec_sig_results$albutein_vs_veh_v2$pert_iname)


intersect(huvec_sig_results$albutein_vs_veh_degs$pert_iname,huvec_sig_results$albutein_vs_veh$pert_iname)

overlap2 <- intersect(queries_against_lincs2$endothelial$albutein_vs_veh_v2$pert_iname, queri)
```



- Significant results per LINCS2 cell-line:

```{r}
full_table_split_by_cell <- full_table %>%
  split(x=., f = .$cell) 


full_table_split_by_cell %>% lapply(
  function(x){
    x$sig %>% table
  }
)

full_table_split_by_cell$HUVEC %>% View
```

### Best cell lines ranked by how many significant queries 

```{r}

n_drugs_tested_per_cellline <- full_table_split_by_cell %>%
  lapply(nrow)

n_drugs_tested_per_cellline <- data.frame(
 cell = names(n_drugs_tested_per_cellline),
 n_drugs_tested = unlist(n_drugs_tested_per_cellline)
)

n_drugs_sig_per_cellline <- full_table_split_by_cell %>%
  lapply(function(x){
    x %>%
      dplyr::filter(sig == TRUE) %>%
      nrow()
  })

n_drugs_sig_per_cellline <- data.frame(
 cell = names(n_drugs_sig_per_cellline),
 n_drugs_sig = unlist(n_drugs_sig_per_cellline)
)

best_cell_lines <- full_join(n_drugs_tested_per_cellline, 
          n_drugs_sig_per_cellline,
          by = "cell") %>% 
  dplyr::mutate(prop = n_drugs_sig / n_drugs_tested) %>%
  dplyr::arrange(desc(prop))

View(best_cell_lines)
```

#### p-value Distribution

- What does the distribution of p-values look like?

```{r}
plots_pval_histogram <- full_table %>%
  ggplot(aes(x = WTCS_Pval)) +
  geom_histogram(bins = 100,
                 fill = "white",
                 color = "black") +
  labs(x = "WTCS p-value", y = "Frequency") +
  ggtitle("CMap WTCS values: p-value histogram")
plots_pval_histogram

ggsave(file.path(output_dir,"pval_histogram.pdf"),
       plots_pval_histogram,
       width=11, height=8, units = "in")
```

#### NCS Distribution

```{r}
plots_NCS_histogram <- full_table %>%
  ggplot(aes(x = NCS)) +
  geom_histogram(bins = 100,
                 fill = "white",
                 color = "black") +
  labs(x = "NCS value", y = "Frequency") +
  ggtitle("CMap NCS values histogram")
plots_NCS_histogram

ggsave(file.path(output_dir,"ncs_histogram.pdf"),
       plots_NCS_histogram,
       width=11, height=8, units = "in")
```


#### NCSct Distribution

```{r}
plots_NCSct_histogram <- full_table %>%
  ggplot(aes(x = NCSct)) +
  geom_histogram(bins = 100,
                 fill = "white",
                 color = "black") +
  labs(x = "NCSct value", y = "Frequency") +
  ggtitle("CMap NCSct values histogram")
plots_NCSct_histogram

ggsave(file.path(output_dir,"ncsct_histogram.pdf"),
       plots_NCSct_histogram,
       width=11, height=8, units = "in")
```

## Over-Representation Analysis (ORA) 

### Drug Classes (MOAss)

- The `MOAss` column in `full_table` is basically the drug class (it is supposed to represent the mechanism of action). 
- We can do an ORA with hypergeometric test to see if any drug classes (MOAss)'s are over-represented. 

```{r}
# Long-form, with each drug class on separate row. 
full_table_long <- full_table %>%
  tidyr::separate_rows(MOAss, sep = ";") %>%
  dplyr::mutate(MOAss = stringr::str_trim(MOAss)) %>%
  dplyr::filter(!is.na(MOAss)) %>%
  split(x=., f = .$cell)

ora_moa <- full_table_long %>%
  lapply(function(x){
    x %>%
      ORAUsingFishersExactTest(groupingCol = "MOAss",
                               test = "hypergeometric")
  })

ora_moa %>% saveRDS(file.path(output_dir, "ora_moa.rds"))
ora_moa %>% 
  lapply(function(x){
    x %>%
      dplyr::group_by(sig) %>%
      dplyr::summarize(n=n())
      }) %>%
  bind_rows(.id="cell")
```


## Over-Represented Indications

- Are there particular indications that are over-represented amongst the significant drugs as compared to all queries?

- First, check how many indication-drug pairs are already present in the results table. 

- In the HUVEC cell line, there are 90 / 393 drugs (23%) from the results that have indication-drug pairs. 

```{r}
nIndicationsInHUVEC <- full_table %>%
  dplyr::filter(cell == "HUVEC") %>%
  .$indication_class %>% unique %>% length

nDrugsInHUVEC <- full_table%>%
  dplyr::filter(cell == "HUVEC") %>%
  .$pert_iname %>% unique %>% length

nDrugIndicationPairsHUVEC <- full_table %>%
  dplyr::filter(cell == "HUVEC") %>%
  dplyr::filter(!is.na(indication_class) & 
                  !is.na(pert_iname)) %>% 
  nrow()
nDrugIndicationPairsHUVEC

nDrugIndicationPairsHUVEC/nDrugsInHUVEC
```

In terms of significant drugs, there are 34 in HUVECs, with 9 of them having an indication. There are only 5 unique indications. 

```{r}
nDrugsInHUVECSig <- full_table %>%
  dplyr::filter(cell == "HUVEC") %>%
  dplyr::filter(sig == TRUE) %>%
  .$pert_iname %>% unique %>% length
nDrugsInHUVECSig

nIndicationsInHUVECSig <- full_table %>%
  dplyr::filter(cell == "HUVEC") %>%
  dplyr::filter(sig == TRUE) %>%
  .$indication_class %>% unique %>% length
nIndicationsInHUVECSig

nDrugIndicationPairsHUVECSig <- full_table %>%
  dplyr::filter(cell == "HUVEC") %>%
  dplyr::filter(sig == TRUE) %>%
  dplyr::filter(!is.na(pert_iname), !is.na(indication_class)) %>%
  nrow()
nDrugIndicationPairsHUVECSig
```

Lets see if we can improve this through the indications from DrugBank. 

```{r}
indications_drugbank <- readRDS(here("data", "indications_full_final_syn.rds")) %>%
  dplyr::mutate(name = tolower(name)) %>%
  dplyr::mutate(synonym = tolower(synonym))
```

- Quick check of how many of the DrugBank synonyms are in CMap. Out of the 339 drugs, there are 240 that have indications from the DrugBank table. 

```{r}
drugsInHuvec <- full_table %>%
  dplyr::filter(cell == "HUVEC") %>%
  .$pert_iname %>% unique %>% tolower() 
length(drugsInHuvec)
table(drugsInHuvec %in% c(indications_drugbank$synonym, indications_drugbank$name))
table(drugsInHuvec %in% indications_drugbank$name)
table(drugsInHuvec %in% indications_drugbank$synonym)
```

#### Create mapping table 

- This sounds good so let's do the join.

```{r}
mappingTableDrugIndication <- full_table %>%
  dplyr::select(pert_iname) %>%
  dplyr::distinct(.keep_all=T) %>%
  left_join(indications_drugbank %>% dplyr::select(-synonym) %>% dplyr::distinct(.keep_all = T), by = 
              c("pert_iname"="name"),
            multiple = "all") %>%
  left_join(indications_drugbank %>% dplyr::select(-name,
                                                   -drug_id,
                                                   -cas_number) %>%
              dplyr::distinct(.keep_all = T),
            by = c("pert_iname"="synonym"),
            multiple = "all") %>%
  dplyr::mutate(indication = case_when(
    !is.na(indication.x) ~ indication.x,
    !is.na(indication.y) ~ indication.y,
    TRUE ~ NA
  )) %>%
  dplyr::select(-indication.x, -indication.y) %>%
  dplyr::distinct(.keep_all=T) %>%
  dplyr::filter(!is.na(indication))


View(mappingTableDrugIndication)

mappingTableDrugIndication %>%
  saveRDS(file.path(output_dir, "mappingTableDrugIndication.rds"))

mappingTable2 <- mappingTableDrugIndication %>%
  dplyr::select(pert_iname, indication) %>%
  tidyr::nest(pert_iname) 
mappingTable2$data %<>% lapply(function(x){
  x %>% unlist %>% unname
})
mappingTable2 %<>%
  dplyr::rename(drugs = data)
View(mappingTable2)
mappingTable2 %>%
  saveRDS(file.path(output_dir, "mappingTable2.rds"))
```

#### Join mapping table to our results

```{r}
full_table2 <- full_table %>%
  left_join(
    mappingTableDrugIndication,
    by = c("pert_iname"),
    multiple = "all") 
                                                          
ora_indication <- ORAUsingFishersExactTest(
  full_table2,
  groupingCol = "indication",
  test = "hypergeometric")

ora_indication %>% saveRDS(file.path(output_dir,
                                     "ora_indication.rds"))

colnames(ora_indication)
ora_indication$sig %>% table

View(ora_indication)
```

## DisGeNet analysis - ORA of targets-indications

### Tidy table

- Retain relevant columns
- Separate out gene symbols, one on each row. 

**Note**: Below, here we are focusing on the HUVEC cell line as this is most comparable to endothelial cells from the transcriptlomics data. 

```{r}
full_table_tidy <- full_table %>%
  dplyr::filter(cell == "HUVEC") %>%
  dplyr::select(pert, pert_iname, 
                starts_with("WTC"),
                starts_with("NCS"),
                MOAss,
                t_gn_sym,
                sig)   %>%
  tidyr::separate_rows(t_gn_sym, sep = ";") %>%
  dplyr::mutate(t_gn_sym =  stringr::str_trim(t_gn_sym)) 
nrow(full_table_tidy)

head(full_table_tidy)
```


### Filter sig results, extract gene symbols

```{r}
sig_res <- full_table_tidy %>%
  dplyr::filter(sig == TRUE) 
  
sig_res %>%
  dplyr::distinct(pert_iname, t_gn_sym) %>% 
  dplyr::group_by(pert_iname) %>%
  dplyr::summarise(n=n()) %>% .$n %>%
  hist(breaks=50)

  sig_res <- sig_res%>% .$t_gn_sym %>% unique()
sig_res
length(sig_res)
```

### Enrichment (over-representation) analysis

- This will simply be performed with `disease_enrichment` function already implemented in DisGeNet's R package. 

```{r}
res_enrich <- disease_enrichment(entities = sig_res)

summaryPlot <- plot( res_enrich, class = "Enrichment", count =3,  cutoff= 0.05, nchars=70)

summaryPlot %>% ggsave(
  file.path(output_dir,
            "summaryPlot_disgenet.pdf"),
  . ,
  width = 11,
  height = 20,
  units = "in"
)

res_table <- res_enrich@qresult %>%
  dplyr::mutate(sig = FDR <= 0.05)
View(res_table)
res_table$sig %>% table

sig_table <- res_table %>%
  dplyr::filter(sig == TRUE)

nrow(sig_table)
nrow(res_table)
```

### Export Results

```{r}
res_table %>% 
  write_csv(file.path(output_dir, 
                      "disgenet_target_indication_ora.csv"))
res_table %>%
  saveRDS(file.path(output_dir, "res_table.rds"))
```
